sw/device/tests/edn_sw_mode.c

```diff
@@ -0,0 +1,187 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "hw/ip/aes/model/aes_modes.h"
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_aes.h"
+#include "sw/device/lib/dif/dif_csrng.h"
+#include "sw/device/lib/dif/dif_edn.h"
+#include "sw/device/lib/dif/dif_entropy_src.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/aes_testutils.h"
+#include "sw/device/lib/testing/edn_testutils.h"
+#include "sw/device/lib/testing/entropy_testutils.h"
+#include "sw/device/lib/testing/rand_testutils.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+
+#include "edn_regs.h"  // Generated
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+#define TIMEOUT (1000 * 1000)
+// AES with Domain-Oriented Masking (DOM) takes 72 cycles per 16B data
+// block for a 256 bit key. Multiply this number by 20, to be on the safe side,
+// to get the number of micro seconds for the timeout.
+#define PLAIN_TEXT_BYTES (4 * 4)
+#define TIMEOUT_AES (72 * (PLAIN_TEXT_BYTES / 16) * 20)
+
+static dif_entropy_src_t entropy_src;
+static dif_csrng_t csrng;
+static dif_edn_t edn0;
+static dif_edn_t edn1;
+static dif_aes_t aes;
+
+OTTF_DEFINE_TEST_CONFIG();
+
+// Initializes the peripherals used in this test.
+static void init_peripherals(void) {
+  CHECK_DIF_OK(dif_entropy_src_init(
+      mmio_region_from_addr(TOP_EARLGREY_ENTROPY_SRC_BASE_ADDR), &entropy_src));
+  CHECK_DIF_OK(dif_csrng_init(
+      mmio_region_from_addr(TOP_EARLGREY_CSRNG_BASE_ADDR), &csrng));
+  CHECK_DIF_OK(
+      dif_edn_init(mmio_region_from_addr(TOP_EARLGREY_EDN0_BASE_ADDR), &edn0));
+  CHECK_DIF_OK(
+      dif_edn_init(mmio_region_from_addr(TOP_EARLGREY_EDN1_BASE_ADDR), &edn1));
+  CHECK_DIF_OK(
+      dif_aes_init(mmio_region_from_addr(TOP_EARLGREY_AES_BASE_ADDR), &aes));
+}
+
+// Wrapper function for the AES_TESTUTILS_WAIT_FOR_STATUS macro.
+static status_t aes_wait_for_status(dif_aes_status_t aes_state, bool flag,
+                                    uint32_t timeout) {
+  AES_TESTUTILS_WAIT_FOR_STATUS(&aes, aes_state, flag, timeout);
+  return OK_STATUS();
+}
+
+// Wrapper function for the DIF_RETURN_IF_ERROR macro. This function returns
+// the EDN status once the EDN is ready.
+static dif_result_t edn1_return_ready_status(dif_edn_status_t dif_edn_status) {
+  bool ready = false;
+  while (!ready) {
+    DIF_RETURN_IF_ERROR(dif_edn_get_status(&edn1, dif_edn_status, &ready));
+  }
+  return kDifOk;
+}
+
+// Generate bits in SW mode.
+static void edn_generate_sw_bits(void) {
+  dif_edn_seed_material_t seed0 = edn_testutils_seed_material_build(true);
+  const dif_edn_entropy_src_toggle_t entropy_src_enable =
+      kDifEdnEntropySrcToggleEnable;
+  CHECK_DIF_OK(dif_edn_instantiate(&edn0, entropy_src_enable, &seed0));
+  CHECK_DIF_OK(dif_edn_generate_start(&edn0, 4095));
```

Comment:
I think you've somewhere defined a constant for the `4095` literal. I suggest using that here. But note that the argument of `dif_edn_generate_start` is the number of 32-bit words. Whereas 4095 is the maximum number of 128-bit blocks.

Reply:
You're right, I'll change it in my next push. Regarding the maximum number of blocks, I did this on purpose. Basically, as you already know, this line just triggers the EDN to just start generating the needed entropy for the AES block to finish operation. I just chose the maximum to be sure that the AES has enough and since the test is over right after AES finishes this doesn't matter much I think.